idx/memory/tag_query.go

```diff
@@ -2,504 +2,240 @@ package memory
+	byId      map[schema.MKey]*interning.ArchiveInterned // the metric index by ID, set by Run()/RunGetTags()
+	startWith int                                        // the expression index to start with
 }
 
 // NewTagQueryContext takes a tag query and wraps it into all the
-// context structs necessary to execute the query on the index
+// context structs necessary to execute the query on the indexes
 func NewTagQueryContext(query tagquery.Query) TagQueryContext {
-	kvsFromExpressions := func(expressions []tagquery.Expression) []kv {
-		res := make([]kv, len(expressions))
-		for i := range expressions {
-			res[i] = kv{Tag: expressions[i].Tag}
-		}
-		return res
+	ctx := TagQueryContext{
+		query:     query,
+		startWith: -1,
 	}
 
-	kvReFromExpression := func(expression tagquery.Expression) kvRe {
-		return kvRe{
-			kv:         kv{Tag: expression.Tag},
-			Regex:      expression.Regex,
-			matchCache: &sync.Map{},
-			missCache:  &sync.Map{},
-		}
-	}
-
-	kvResFromExpressions := func(expressions []tagquery.Expression) []kvRe {
-		res := make([]kvRe, len(expressions))
-		for i := range expressions {
-			res[i] = kvReFromExpression(expressions[i])
-		}
-		return res
-	}
-
-	return TagQueryContext{
-		wg:        &sync.WaitGroup{},
-		equal:     kvsFromExpressions(query.Expressions[tagquery.EQUAL]),
-		match:     kvResFromExpressions(query.Expressions[tagquery.MATCH]),
-		notEqual:  kvsFromExpressions(query.Expressions[tagquery.NOT_EQUAL]),
-		notMatch:  kvResFromExpressions(query.Expressions[tagquery.NOT_MATCH]),
-		prefix:    kvsFromExpressions(query.Expressions[tagquery.PREFIX]),
-		tagClause: query.TagClause,
-		tagPrefix: query.TagPrefix,
-		tagMatch:  kvReFromExpression(query.TagMatch),
-		startWith: query.StartWith,
-		from:      query.From,
-	}
+	return ctx
 }
 
-// getInitialByEqual generates the initial resultset by executing the given equal expression
-func (q *TagQueryContext) getInitialByEqual(expr kv, idCh chan schema.MKey, stopCh chan struct{}) {
-	defer q.wg.Done()
-
-KEYS:
-	for k := range q.index[expr.Key][expr.Value] {
-		select {
-		case <-stopCh:
-			break KEYS
-		case idCh <- k:
-		}
+func (q *TagQueryContext) prepareExpressions(idx TagIndex) {
+	type expressionCost struct {
+		cost          uint32
+		expressionIdx int
 	}
+	costs := make([]expressionCost, len(q.query.Expressions))
 
-	close(idCh)
-}
+	for i, expr := range q.query.Expressions {
+		costs[i].expressionIdx = i
 
-// getInitialByPrefix generates the initial resultset by executing the given prefix match expression
-func (q *TagQueryContext) getInitialByPrefix(expr kv, idCh chan schema.MKey, stopCh chan struct{}) {
-	defer q.wg.Done()
-
-VALUES:
-	for v, ids := range q.index[expr.Key] {
-		if !strings.HasPrefix(v, expr.Value) {
+		keyPtr, found := expr.GetKeyPtr()
+		if !found {
+			// the key of this expression has not been interned, this means it cannot exist in the index.
+			// so this expression will be cheap to evaluate
```

Comment:
example? maybe as a comment

Reply:
added

Reply:
thanks. the new comment makes a lot of sense.
are there any expression types for which this optimization is not true?

Reply:
In all cases where an expression's `.OperatesOnTag()` returns `false`, the return value of `.GetKey()` has to match the tag value in the index exactly. So if the return value of `.GetKey()` cannot be found in the interning layer, then we know that this tag is not present in the index.